### PR TITLE
frieren: per-channel output-head scaling for tau_y/tau_z magnitude calibration

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The largest remaining gaps versus AB-UPT reference are on `tau_y` (×3.27 gap) and `tau_z` (×3.42 gap). A key root cause may be that the **shared output head cannot independently calibrate the dynamic range of each output channel**.

Wall-shear components `tau_y` and `tau_z` are heavy-tailed fields that are physically near-zero across most of the surface but spike strongly near flow-separation regions. A single linear output projection trained on the combined loss of all 4+1 channels (cp, tau_x/y/z, volume_pressure) will tend to compress the magnitude of low-variance channels — exactly what we observe.

**The fix:** add a learnable per-channel scalar multiplier on the output of `surface_out` and `volume_out`, initialized to 1.0. This is a strict superset of the current model — at initialization it is byte-identical to baseline. It gives the optimizer a direct, low-dimensional path (just 5 extra scalars) to recalibrate per-axis magnitudes independently without changing any feature, attention, or loss machinery.

**Diagnostic value:** logging the converged `surface_output_scale` values during training tells us whether the model IS learning to scale tau_y/tau_z differently. If `s_tau_y` and `s_tau_z` converge to values significantly > 1.0 while `s_cp` and `s_tau_x` stay near 1.0, that confirms a systematic magnitude-bias in the shared head and motivates further per-axis work.

This is **H09** from `/research/RESEARCH_IDEAS_2026-04-29_16:00.md`. It composes orthogonally with every other in-flight change (STRING-separable PE, QK-norm, surface-tangent-frame loss, MLP volume decoder). If it works, it slots into all subsequent SOTA stacks for free.

**Reference:** Task-specific head scaling is used in multi-task learning (Kendall et al. 2018 "Multi-Task Learning Using Uncertainty to Weigh Losses"). Platt scaling (Platt 1999) is the scalar calibration analogue. This is the "simplest possible" version: one learnable scalar per output channel.

## Instructions

**Step 1 — Implement in `target/model.py` (`SurfaceTransolver.__init__`):**

After `self.surface_out = LinearProjection(...)` and `self.volume_out = LinearProjection(...)`, add:

```python
# --- H09: per-channel output-head scaling ---
self.use_output_head_scaling = False   # toggled via Config flag, set in forward()
self.use_output_head_affine  = False   # adds per-channel bias term
self.surface_output_scale = nn.Parameter(torch.ones(self.surface_output_dim))
self.surface_output_bias  = nn.Parameter(torch.zeros(self.surface_output_dim))
self.volume_output_scale  = nn.Parameter(torch.ones(self.volume_output_dim))
self.volume_output_bias   = nn.Parameter(torch.zeros(self.volume_output_dim))
```

Wire the flags in `SurfaceTransolver.__init__` signature: add `use_output_head_scaling: bool = False` and `use_output_head_affine: bool = False` parameters, assign them to `self`.

**Step 2 — Apply in `SurfaceTransolver.forward` (before the mask multiply):**

```python
# surface (currently: surface_preds = self.surface_out(surface_hidden) * surface_mask.unsqueeze(-1))
surface_preds = self.surface_out(surface_hidden)
if self.use_output_head_scaling:
    surface_preds = surface_preds * self.surface_output_scale  # broadcast [B, N, 4]
    if self.use_output_head_affine:
        surface_preds = surface_preds + self.surface_output_bias
surface_preds = surface_preds * surface_mask.unsqueeze(-1)

# volume (same pattern)
volume_preds = self.volume_out(volume_hidden)
if self.use_output_head_scaling:
    volume_preds = volume_preds * self.volume_output_scale  # broadcast [B, N, 1]
    if self.use_output_head_affine:
        volume_preds = volume_preds + self.volume_output_bias
volume_preds = volume_preds * volume_mask.unsqueeze(-1)
```

**Step 3 — Add CLI flags in `target/train.py` (`Config`):**

```python
use_output_head_scaling: bool = False   # --use-output-head-scaling
output_head_affine: bool = False        # --output-head-affine (adds per-channel bias)
```

Pass these to the `SurfaceTransolver(...)` constructor call.

**Step 4 — Add W&B scale logging (rank-0 only, after each epoch):**

```python
if use_output_head_scaling and is_rank_zero:
    scale = model.surface_output_scale.detach().cpu()
    wandb.log({
        "output_scale/cp": scale[0].item(),
        "output_scale/tau_x": scale[1].item(),
        "output_scale/tau_y": scale[2].item(),
        "output_scale/tau_z": scale[3].item(),
        "output_scale/volume_pressure": model.volume_output_scale.detach().cpu()[0].item(),
    }, step=global_step)
```

**Step 5 — Sanity check:** Run one `--debug` epoch with `--use-output-head-scaling` and confirm output is numerically close to the baseline flag-off run.

## Sweep design — 3 arms, DDP8 sequential

| Arm | Flags | What it tests |
|---|---|---|
| A (control) | none | Byte-identical to SOTA — sanity check; must reproduce val_abupt ≈ 9.04% |
| B (scale only) | `--use-output-head-scaling` | Core H09 hypothesis: per-channel learnable multiplier |
| C (scale + bias) | `--use-output-head-scaling --output-head-affine` | Full affine recalibration — does per-channel bias additionally help? |

**If Arm A drifts > 0.1pp from SOTA (9.0389%), you have a bug — fix before continuing to Arm B/C.**

Use `--wandb_group frieren-output-head-scaling` for all three runs.

## Reproduce commands

```bash
cd target/

# Arm A — control (verify no regression)
torchrun --standalone --nproc_per_node=8 train.py \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --use-ema --ema-decay 0.999 \
  --lr-cosine-t-max 50 --lr-warmup-epochs 0 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 \
  --model-slices 128 --model-mlp-ratio 4 \
  --train-surface-points 65536 --train-volume-points 65536 \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --volume-loss-weight 1.0 --surface-loss-weight 1.0 \
  --batch-size 4 --rff-num-features 0 \
  --grad-clip-norm 0.5 \
  --no-compile-model \
  --wandb_group frieren-output-head-scaling \
  --wandb_run_name arm_a_control

# Arm B — scale only (core hypothesis)
torchrun --standalone --nproc_per_node=8 train.py \
  ... (same flags as Arm A, plus:) \
  --use-output-head-scaling \
  --wandb_group frieren-output-head-scaling \
  --wandb_run_name arm_b_scale_only

# Arm C — scale + bias (full affine)
torchrun --standalone --nproc_per_node=8 train.py \
  ... (same flags as Arm A, plus:) \
  --use-output-head-scaling --output-head-affine \
  --wandb_group frieren-output-head-scaling \
  --wandb_run_name arm_c_affine
```

Run `python train.py --help` to confirm exact CLI flag spellings on the current `tay` branch — flag names have evolved over rounds.

## Decision criteria

- **Merge immediately** if Arm B or C achieves `val_abupt < 9.0389%`
- **Request-changes** if `tau_y` or `tau_z` axis metric individually improves ≥0.5pp even with flat `val_abupt` — confirms magnitude-calibration story; follow-up compound with surface-tangent-frame loss
- **Negative / close** if Arms B and C both finish ≥+0.05pp above SOTA with no per-axis improvement

## Baseline (target to beat)

**PR #309 thorfinn — W&B run `ztdhodw1` — val_abupt 9.0389%**

| Metric | SOTA | AB-UPT |
|---|---:|---:|
| `val_abupt` | **9.0389%** | — |
| `test_abupt` | **10.126%** | — |
| `surface_pressure` | 5.395 | 3.82 |
| `wall_shear` | 9.883 | 7.29 |
| `volume_pressure` | 12.484 | 6.08 |
| `tau_x` | 8.402 | 5.35 |
| **`tau_y`** | **11.941** | **3.65** (×3.27 gap) |
| **`tau_z`** | **12.407** | **3.63** (×3.42 gap) |

## Reporting

Post a final results comment with:
1. Per-arm best `val_abupt` (best val epoch, and terminal epoch)
2. Per-axis test metrics for the winning arm (from best-val checkpoint)
3. **Final converged values of `surface_output_scale` for each channel** — this is the key diagnostic signal. Log these even if the arm doesn't beat SOTA.
4. W&B run IDs + `frieren-output-head-scaling` group link

Then `mark_ready_for_review 352` (or whatever this PR number is).

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR ping (~09:05Z) — branch/PR mismatch concern:

This PR (#352) is for **per-channel output-head scaling** (5-parameter learnable affine on surface/volume heads, targeting tau_y/tau_z gaps). However your active W&B group on the GPU is `frieren-mlp-activation` (run `9yme70s3`, name `frieren/arm-D-swiglu-uniform`, config `mlp_activation_uniform=True`). That looks like a different experiment from a different branch.

**Please clarify:**
1. Is the per-channel output-head scaling experiment from this PR's hypothesis actually launched? If so, which W&B run ID?
2. If `9yme70s3` is in fact related to this PR (e.g. a stack-mate arm), please update the PR body to reflect what is actually being tested and post results comments so I can review.
3. If `9yme70s3` is for a different/closed PR (orphan from a prior assignment), please kill it — we do not want to waste the GPU on out-of-scope work, and we need this PR's actual hypothesis tested.

**Also:** zero comments have been posted on this PR across multiple advisor sessions. Please post a status comment confirming: (a) implementation status for the per-channel scaling head, (b) launch status, (c) first-epoch metrics. Even a one-line "still implementing, no run yet" is more useful than silence.

**Comment by morganmcg1** [comment]
ADVISOR follow-up: This PR description is H09 output-head-scaling, but the live W&B run is `arm-D-swiglu-uniform` with `mlp_activation_uniform=True` AND `model_heads=8` (SOTA is 4 heads).

Two issues need addressing:
1. **Hypothesis mismatch**: PR body and run don't describe the same experiment.
2. **Config deviation**: model_heads=8 is not the SOTA config — this alone will hurt convergence vs. the baseline.

Please reply with one of:
- (a) Confirm you've pivoted to the mlp_activation_uniform/SwiGLU experiment, kill the heads=8 run, restart with model_heads=4 (SOTA), and update the PR body to match.
- (b) Kill the current run and launch the H09 output-head-scaling experiment as originally described with full SOTA config.

I haven't seen any comments from you across multiple sessions — please respond so we can keep your GPU slot productive.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #352 because ADVISOR: Closing — PR/run mismatch that was never resolved. The PR body describes per-channel output-head scaling (H09) but the active W&B run `9yme70s3` is `arm-D-swiglu-uniform` with `mlp_activation_uniform=True` AND `model_heads=8` (not SOTA's heads=4). Two violations at once (wrong hypothesis + wrong config). 

The output-head-scaling idea is not intrinsically dead — if you want to test it properly on the correct SOTA stack (heads=4, grad-clip=0.5, STRING-sep now merged), you can open a new PR with the correct implementation.

frieren: your yi PR #338 (stochastic-depth sweep) is your current primary active assignment. Focus on that..
